packages/neo-one-monitor/src/BrowserMetricsFactory.js

```diff
@@ -0,0 +1,149 @@
+/* @flow */
+import { utils } from '@neo-one/utils';
+
+import type {
+  MetricsFactory,
+  MetricLabels,
+  MetricConstruct,
+} from './MonitorBase';
+
+import type {
+  CounterMetric,
+  GaugeMetric,
+  HistogramMetric,
+  SummaryMetric,
+} from './types';
+
+let metrics = {
+  counters: {},
+  histograms: {},
+};
+
+export const resetMetricsForTesting = (): void => {
+  metrics = {
+    counters: {},
+    histograms: {},
+  };
+};
+
+export type MetricValue = {|
+  labels?: MetricLabels,
+  count?: number,
+|};
+
+export class CollectingMetric {
+  metric: MetricConstruct;
+  values: Array<MetricValue>;
+
+  constructor(metric: MetricConstruct) {
+    this.metric = metric;
+    this.values = [];
+  }
+
+  reset(): void {
+    this.values = [];
+  }
+}
+
+export type MetricCollection = {|
+  counters: { [name: string]: CollectingMetric },
+  histograms: { [name: string]: CollectingMetric },
+|};
+
+class BrowserCounter extends CollectingMetric implements CounterMetric {
+  inc(labels?: MetricLabels, count?: number): void {
+    this.values.push({ labels, count });
+  }
+}
+
+export class BrowserGauge extends CollectingMetric implements GaugeMetric {
+  // eslint-disable-next-line
+  inc(labels?: MetricLabels, count?: number): void {}
+
+  // eslint-disable-next-line
+  dec(labels?: MetricLabels, count?: number): void {}
+
+  // eslint-disable-next-line
+  set(labels?: MetricLabels, count?: number): void {}
+}
+
+class BrowserHistogram extends CollectingMetric implements HistogramMetric {
+  observe(labels?: MetricLabels, count?: number): void {
+    this.values.push({ labels, count });
+  }
+}
+
+export class BrowserSummary extends CollectingMetric implements SummaryMetric {
+  // eslint-disable-next-line
+  observe(labels?: MetricLabels, count?: number): void {}
+}
+
+export default class BrowserMetricsFactory implements MetricsFactory {
+  collect(): MetricCollection {
+    const currentMetrics = this._serializeJSON(metrics);
+    utils
+      .keys(metrics)
+      .map(metricType =>
+        utils
+          .keys(metrics[metricType])
+          .map(name => metrics[metricType][name].reset()),
+      );
+    return currentMetrics;
+  }
+
+  createGauge(options: MetricConstruct): GaugeMetric {
+    return new BrowserGauge(options);
+  }
+
+  createCounter(options: MetricConstruct): CounterMetric {
+    const counter = new BrowserCounter(options);
+    metrics.counters[counter.metric.name] = counter;
+
+    return counter;
+  }
+
+  createHistogram(options: MetricConstruct): HistogramMetric {
+    const histogram = new BrowserHistogram(options);
+    metrics.histograms[histogram.metric.name] = histogram;
+
+    return histogram;
+  }
+
+  createSummary(options: MetricConstruct): SummaryMetric {
+    return new BrowserSummary(options);
+  }
+
+  _serializeJSON(currentMetrics: MetricCollection): MetricCollection {
+    const copy = utils
```

Comment:
This should be something like
```
return utils.entries(currentMetrics).reduce(
  (acc, [metricType, metricTypeMetrics]) => ({
    ...acc,
    [metricType]: utils.entries(metricTypeMetrics).reduce(
      (innerAcc, [name, metric]) => ({
        ...innerAcc
        [name]: ...
      }),
      {},
    ),
  }),
  {},
);
```

But anyways, I'm not sure what the point of this is anyways, why are we doing a deep clone of `MetricCollection`?

Reply:
Oh I think I see now, you're converting all of the class instances to just the objects. I would do something like `[name]: metric.toJSON()` in the above example code to make that clearer and do the actual conversion inside the `Metric` class. It makes more sense from an OO perspective anyways, why should `MetricsFactory` know how to serialize a particular `Metric`? Instead, the `Metric` should know how to serialize itself.